pkg/gentrait/securityevent/model.go

```diff
@@ -0,0 +1,166 @@
+package securityevent
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/smart-core-os/sc-api/go/types"
+	"github.com/smart-core-os/sc-golang/pkg/resource"
+	"github.com/vanti-dev/sc-bos/pkg/gen"
+)
+
+var subSystems = [2]string{"access control", "cctv"}
+
+type Model struct {
+	mu                sync.Mutex // guards allSecurityEvents and genId
+	allSecurityEvents []*gen.SecurityEvent
+	genId             int
+
+	lastSecurityEvent *resource.Value // of *gen.SecurityEvent
+}
+
+func NewModel(opts ...resource.Option) *Model {
+	defaultOpts := []resource.Option{resource.WithInitialValue(&gen.SecurityEvent{})}
+	opts = append(defaultOpts, opts...)
+
+	m := &Model{
+		lastSecurityEvent: resource.NewValue(opts...),
+	}
+
+	// let's add some events to start with so we can test the list method without waiting
+	startTime := time.Now().Add(-100 * time.Minute)
+	for i := 0; i < 100; i++ {
+		_, _ = m.GenerateSecurityEvent(timestamppb.New(startTime))
+		startTime = startTime.Add(time.Minute)
+	}
+
+	return m
+}
+
+// AddSecurityEvent manually add a security event to the model
+func (m *Model) AddSecurityEvent(se *gen.SecurityEvent, opts ...resource.WriteOption) (*gen.SecurityEvent, error) {
+	v, err := m.lastSecurityEvent.Set(se, opts...)
+	if err != nil {
+		return nil, err
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.allSecurityEvents = append(m.allSecurityEvents, se)
+	return v.(*gen.SecurityEvent), nil
+}
+
+// GenerateSecurityEvent generates a security event and adds it to the model
+func (m *Model) GenerateSecurityEvent(time *timestamppb.Timestamp) (*gen.SecurityEvent, error) {
+	m.mu.Lock()
+	se := &gen.SecurityEvent{
+		SecurityEventTime: time,
+		Description:       "Generated security event",
+		Id:                strconv.Itoa(m.genId),
+		Source: &gen.SecurityEvent_Source{
+			Subsystem: subSystems[m.genId%2],
+		},
+		EventType: gen.SecurityEvent_EventType(m.genId % 22),
+	}
+	m.genId++
+	m.mu.Unlock()
+	return m.AddSecurityEvent(se)
+}
+
+func (m *Model) ListSecurityEvents(req *gen.ListSecurityEventsRequest) (*gen.ListSecurityEventsResponse, error) {
```

Comment:
All other models we have try not to require the rpc request/response objects to interact with the data. Could you add a `ListSecurityEvents(start, end int) []*gen.SecurityEvent` function? Would that work?